### PR TITLE
Partial index 2.0

### PR DESCRIPTION
**Pull request by ashtul**
Partial indexing by removing (not indexing) any field which does not fit the schema.

### CODE REVIEW

File: src/document.c [hunk 1]
```diff
@@ -573,8 +575,16 @@ int Document_AddToIndexes(RSAddDocumentCtx *aCtx) {
...
       PreprocessorFunc pp = preprocessorMap[ii];
       if (pp(aCtx, &doc->fields[i], fs, fdata, &aCtx->status) != 0) {
-        ourRv = REDISMODULE_ERR;
-        goto cleanup;
+        // field failed to index. we remove the field from doc and continue
+        uint32_t lastField = doc->numFields - 1;
+        if (i < lastField) {
+          DocumentField tmp = doc->fields[i];
+          doc->fields[i] = doc->fields[lastField];
+          doc->fields[lastField] = tmp;
```
Comment (rafie) [inline src/document.c:583]: We replace field [i] with [lastField], thus dropping field [i] from the list, right? So why the swap?

Comment (ashtul) [reply src/document.c:583]: because the content is being freed in Document_clear (I believe).

Comment (rafie) [reply src/document.c:583]: So in Document_clear we do the following:
```
    for (size_t ii = 0; ii < d->numFields; ++ii) {
      if (d->flags & DOCUMENT_F_OWNSTRINGS) {
        rm_free((void *)d->fields[ii].name);
      }
      if (d->fields[ii].text) {
        RedisModule_FreeString(RSDummyContext, d->fields[ii].text);
      }
    }
```
But here, we doing `--doc->numFields`, thus pulling [lastField] out of range.
I'm sorry if I'm bugging you, but something feels fishy here.

Comment (ashtul) [reply src/document.c:583]: Defiantly is.
In one of the iterations, fields were freed here but then retry to free again later.

Comment (ashtul) [reply src/document.c:583]: This behavior will be fixed with PR #1462 as we will have one copy of document

**Comment by oshadmi** [comment]
Stale

**Pull request closed without merging**